jobsdb/jobsdb.go

```diff
@@ -1863,66 +1864,73 @@ func (jd *Handle) GetToProcess(ctx context.Context, params GetQueryParams, more
 
 var cacheParameterFilters = []string{"source_id", "destination_id"}
 
-func (jd *Handle) GetPileUpCounts(ctx context.Context) (map[string]map[string]int, error) {
+func (jd *Handle) GetPileUpCounts(ctx context.Context) error {
 	if !jd.dsMigrationLock.RTryLockWithCtx(ctx) {
-		return nil, fmt.Errorf("could not acquire a migration read lock: %w", ctx.Err())
+		return fmt.Errorf("could not acquire a migration read lock: %w", ctx.Err())
 	}
 	defer jd.dsMigrationLock.RUnlock()
 	if !jd.dsListLock.RTryLockWithCtx(ctx) {
-		return nil, fmt.Errorf("could not acquire a dslist read lock: %w", ctx.Err())
+		return fmt.Errorf("could not acquire a dslist read lock: %w", ctx.Err())
 	}
 	dsList := jd.getDSList()
 	jd.dsListLock.RUnlock()
-	statMap := make(map[string]map[string]int)
 
-	for _, ds := range dsList {
-		queryString := fmt.Sprintf(`with joined as (
-			select
-			  j.job_id as jobID,
-			  j.custom_val as customVal,
-			  s.id as statusID,
-			  s.job_state as jobState,
-			  j.workspace_id as workspace
-			from
-			  %[1]q j
-			  left join "v_last_%[2]s" s on j.job_id = s.job_id
-			where (
-			  s.job_state not in ('aborted', 'succeeded', 'migrated')
-			  or s.job_id is null
-			)
-		  )
-		  select
-			count(*),
-			customVal,
-			workspace
-		  from
-			joined
-		  group by
-			customVal,
-			workspace;`, ds.JobTable, ds.JobStatusTable)
-		rows, err := jd.dbHandle.QueryContext(ctx, queryString)
-		if err != nil {
-			return nil, err
-		}
+	queryString := `with joined as (
+		select
+		  j.job_id as jobID,
+		  j.custom_val as customVal,
+		  s.id as statusID,
+		  s.job_state as jobState,
+		  j.workspace_id as workspace
+		from
+		  %[1]q j
+		  left join "v_last_%[2]s" s on j.job_id = s.job_id
+		where (
+		  s.job_state not in ('aborted', 'succeeded', 'migrated')
+		  or s.job_id is null
+		)
+	  )
+	  select
+		count(*),
+		customVal,
+		workspace
+	  from
+		joined
+	  group by
+		customVal,
+		workspace;`
```

Comment:
I haven't checked but we should have an index in `job_state`, correct? Just trying to understand if we are leveraging indexes properly here. Mostly because before we were running these sequentially whilst now we're issuing X of these queries in parallel so I think it's even more important to make sure that indexing is helping here.

Reply:
```
QUERY PLAN
 --- 
 HashAggregate  (cost=19679.89..19679.94 rows=5 width=41) 
   Group Key: j.custom_val, j.workspace_id 
   ->  Hash Right Join  (cost=8774.07..19267.59 rows=54974 width=33) 
         Hash Cond: (rt_job_status_104819.job_id = j.job_id) 
         Filter: (((rt_job_status_104819.job_state)::text <> ALL ('{aborted,succeeded,migrated}'::text[])) OR (rt_job_status_104819.job_id IS NULL)) 
         ->  Unique  (cost=0.42..9795.26 rows=55339 width=189) 
               ->  Index Scan using idx_rt_job_status_104819_jid_id on rt_job_status_104819  (cost=0.42..9518.18 rows=110832 width=189) 
         ->  Hash  (cost=8076.07..8076.07 rows=55807 width=41) 
               ->  Seq Scan on rt_jobs_104819 j  (cost=0.00..8076.07 rows=55807 width=41) 

```
The last status view uses one index on the status table.
no index for `job_state` being used though, will dig deeper into this.

Reply:
Have given it some thought over the day. But what's in use looks like the best way to get what we want.
The same pattern is user everywhere(all routers and other consumers interested in jobs of different states).
Filtering by job_state **_after_** the join is by design as we don't want to query extra jobs.

And as for running them in parallel, I think it should be safe because of the same pattern being used by many routers at once.

Reply:
OK cool, no problem. It's configurable anyway so worst case if something happens we'll set the concurrency to 1 and it'll work like before :+1: